src/extensions/agent_modules/shared/state_store/src/agent_state_store.c

```diff
@@ -62,379 +68,65 @@ static char* SafeStrdup(const char* str)
-        Log_Error("Invalid input (key=%p, out_value=%p).", key, out_value);
-        return ADUC_STATE_STORE_RESULT_ERROR;
-    }
-
-    ADUC_STATE_STORE_RESULT res = ADUC_STATE_STORE_RESULT_ERROR;
-    sem_wait(&state_semaphore);
-    JSON_Object* root_object = GetRootObject(durable);
-    double result = json_object_dotget_number(root_object, key);
-
-    if (json_object_dotget_value(root_object, key) == NULL
-        || json_value_get_type(json_object_dotget_value(root_object, key)) != JSONNumber)
-    {
-        Log_Warn("Failed to retrieve int value for key %s.", key);
-        goto done;
-    }
-
-    *out_value = (int)result;
-    res = ADUC_STATE_STORE_RESULT_OK;
-done:
-    sem_post(&state_semaphore);
-    Log_Info("Successfully retrieved int value for key %s.", key);
-    return res;
-}
-
-ADUC_STATE_STORE_RESULT ADUC_StateStore_SetInt(bool durable, const char* key, int value)
-{
-    return ADUC_StateStore_SetData(durable, key, json_value_init_number(value));
-}
-
-// Unsigned integer data retrieval and setting
-ADUC_STATE_STORE_RESULT ADUC_StateStore_GetUnsignedInt(bool durable, const char* key, unsigned int* out_value)
-{
-    int result;
-    int status = ADUC_StateStore_GetInt(durable, key, &result);
-    if (status == ADUC_STATE_STORE_RESULT_OK && result >= 0)
-    {
-        *out_value = (unsigned int)result;
-        Log_Info("Successfully retrieved unsigned int value for key %s.", key);
-        return ADUC_STATE_STORE_RESULT_OK;
-    }
-    Log_Warn("Failed to retrieve unsigned int value for key %s.", key);
-    return ADUC_STATE_STORE_RESULT_ERROR;
-}
-
-ADUC_STATE_STORE_RESULT ADUC_StateStore_SetUnsignedInt(bool durable, const char* key, unsigned int value)
-{
-    return ADUC_StateStore_SetInt(durable, key, (int)value); // Note: Assumes the value is within the range of 'int'
-}
-
-ADUC_STATE_STORE_RESULT ADUC_StateStore_GetString(bool durable, const char* key, char** out_value)
-{
-    if (IsNullOrEmpty(key) || out_value == NULL)
+    if (isInitialized)
     {
-        Log_Error("Invalid input (key=%p, out_value=%p).", key, out_value);
-        return ADUC_STATE_STORE_RESULT_ERROR;
-    }
+        free(state.deviceId);
+        free(state.externalDeviceId);
+        free(state.mqttBrokerHostname);
+        free(state.scopeId);
 
-    ADUC_STATE_STORE_RESULT res = ADUC_STATE_STORE_RESULT_ERROR;
-    sem_wait(&state_semaphore);
-    JSON_Object* root_object = GetRootObject(durable);
-    const char* result = json_object_dotget_string(root_object, key);
-    if (result == NULL)
-    {
-        Log_Warn("Failed to retrieve string value for key %s.", key);
-        goto done;
-    }
+        Log_Info("State store terminated successfully.");
+        isInitialized = false;
 
-    *out_value = SafeStrdup(result); // Callers are responsible for freeing this memory.
-    if (*out_value == NULL)
-    {
-        Log_Error("Failed to duplicate string value for key %s.", key);
-        goto done;
+        sem_destroy(&state_semaphore);
```

Comment:
Do we use this in counting or binary? Does this protect the actual state variable? We should acquire the semaphore at the beginning and have a way to block any operations after assuming this is meant for shared memory.

Reply:
By design. No need to acquire the semaphore becaue this will only be called after we exit the loop that was calling the doWork on the agent modules so nothing and also having stopped the worker threads, so it would not be competing with any other thread. For the same reason, no need to block operations after. It should probably just be a mutex and will change it to that later.